src/Compilers/Core/Portable/CodeGen/MethodBody.cs

```diff
@@ -102,7 +105,7 @@ private static ImmutableArray<Cci.SequencePoint> GetSequencePoints(SequencePoint
 
         ImmutableArray<Cci.ExceptionHandlerRegion> Cci.IMethodBody.ExceptionRegions => _exceptionHandlers;
 
-        bool Cci.IMethodBody.LocalsAreZeroed => true;
+        bool Cci.IMethodBody.LocalsAreZeroed => _localsAreZeroed;
```

Comment:
Now that this may vary, it should be another key in the method body cache. Just having same IL is not enough to say that methods may share the body.
Look at `SerializeMethodBody`

Reply:
I wonder how this could be tested. We need to make sure the caching is not interfering. 

Perhaps we can use code that _relies_ on the zero init happening or not happening (probably using stackalloc or "out" parameters in VB), then we may just run bunch of similar methods that only differ in this bit and see if results are as expected?

---
In reply to: [167321124](https://github.com/dotnet/roslyn/pull/24723#discussion_r167321124) [](ancestors = 167321124)

Reply:
This is a good point, but I haven't tested to see whether it's likely that you get non-zero memory when running stackalloc. Our tests would have a dependency on having non-zero random memory, wouldn't they?

Reply:
I was talking to @agocke about this. We use a body from the cache if we can guarantee that localSignatureHandleOpt.IsNil returns true. localSignatureHandleOpt is obtained from `SerializeLocalVariablesSignature`, which seems to return a default object only if there are no locals. Maybe adding another check is not necessary. I still don't know how to test it, though.

Reply:
The AreLocalsZeroed bool is now being used during IL mapping in cache. A test was added to check if method bodies that differ only by SkipLocalsInitAttribute have different Relative Virtual Addresses. The test was failing before the changes in the cache mapping and it's working now.